In [1]:
import tensorflow as tf
import chess
from chess.pgn import read_game
import numpy as np
from anytree import Node
from network import ChessNeuralNetwork
from game import Chess

import time
from agents.nn_agent import NeuralNetworkAgent

In [2]:
def board_generator():
    pgn = open("millionbase-2.22.pgn")
    while True:
        game = read_game(pgn)
        if game and len(list(game.main_line()))>0:
            node = game
            move_number = np.random.randint(0, high=len(list(game.main_line())) - 1) #don't take the last move
            while 2*(node.board().fullmove_number-1)+int(not node.board().turn) < move_number:
                next_node = node.variation(0)
                node = next_node
            yield node.board()
        else:
            pgn.seek(0)

In [3]:
def simple_value(board):
    values = [1, 3, 3, 5, 9]
    s = 0
    for i, v in enumerate(values):
        s += Chess.pad_bitmask(board.pieces_mask(i + 1, 1)).sum() * v
        s -= Chess.pad_bitmask(board.pieces_mask(i + 1, 0)).sum() * v
    return np.tanh(s / 5)

In [4]:
def alpha_beta(node, depth, alpha, beta):
    if depth == 0 or node.board.is_game_over():
        return simple_value(node.board), node

    legal_moves = list(node.board.legal_moves)
    child_boards = [node.board.copy() for _ in legal_moves]
    children = []
    for idx in range(len(node.board.legal_moves)):
        child_boards[idx].push(legal_moves[idx])
        child = Node(str(legal_moves[idx]), parent=node, board=child_boards[idx], move=legal_moves[idx])
        children.append(child)
    n = node
    if node.board.turn:
        v = -100000
        for child in children:
            vv, nn = alpha_beta(child, depth - 1, alpha, beta)
            if vv > v:
                v = vv
                n = nn
            alpha = max(alpha, v)
            if beta <= alpha:
                break  # (* β cut-off *)
        return v, n
    else:
        v = 100000
        for child in children:
            vv, nn = alpha_beta(child, depth - 1, alpha, beta)
            if vv < v:
                v = vv
                n = nn
            beta = min(beta, v)
            if beta <= alpha:
                break  # (* α cut-off *)
        return v, n

In [6]:
sess = tf.Session()
network = ChessNeuralNetwork()
agent = NeuralNetworkAgent('master', network, create_trainer=False)
sess.run(tf.global_variables_initializer())

In [13]:
for m in b.move_stack

d2d4
g8f6
c2c4
g7g6
b1c3
d7d5
c1f4


In [39]:
gen = board_generator()
board = gen.__next__()

def get_leaves(board, depth):
    print(depth)
    if depth==0 or len(list(board.legal_moves))==0:
        print('here')
        return board
        
    legal_moves = list(board.legal_moves)
    child_boards = [board.copy() for _ in legal_moves]
    for i, move in enumerate(legal_moves):
        child_boards[i].push(legal_moves[i])
        
    for leaf in get_leaves(child_boards[i], depth - 1):
         yield leaf
                
list(get_leaves(board, 2))

2
1
0
here
-1
-2
-3
-4
-5
-6
-7
-8
-9
-10
-11
-12
-13
-14
-15
-16
-17
-18
-19
-20
-21
-22
-23
-24
-25
-26
-27
-28
-29
-30
-31
-32
-33
-34
-35
-36
-37
-38
-39
-40
-41
-42
-43
-44
-45
-46
-47
-48
-49
-50
-51
-52
-53
-54
-55
-56
-57
-58
-59
-60
-61
-62
-63
-64
-65
-66
-67
-68
-69
-70
-71
-72
-73
-74
-75
-76
-77
-78
-79
-80
-81
-82
-83
-84
-85
-86
-87
-88
-89
-90
-91
-92
-93
-94
-95
-96
-97
-98
-99
-100
-101
-102
-103
-104
-105
-106
-107
-108
-109
-110
-111
-112
-113
-114
-115
-116
-117
-118
-119
-120
-121
-122
-123
-124
-125
-126
-127
-128
-129
-130
-131
-132
-133
-134
-135
-136
-137
-138
-139
-140
-141
-142
-143
-144
-145
-146
-147
-148
-149
-150
-151
-152
-153
-154
-155
-156
-157
-158
-159
-160
-161
-162
-163
-164
-165
-166
-167
-168
-169
-170
-171
-172
-173
-174
-175
-176
-177
-178
-179
-180
-181
-182
-183
-184
-185
-186
-187
-188
-189
-190
-191
-192
-193
-194
-195
-196
-197
-198
-199
-200
-201
-202
-203
-204
-205
-206
-207
-208
-209
-210
-211
-212
-213
-214
-215
-216
-217
-218
-219
-2

ERROR:root:Internal Python error in the inspect module.
Below is the traceback from this internal error.



Traceback (most recent call last):
  File "/Users/adam/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-39-b945b1917f3f>", line 18, in <module>
    list(get_leaves(board, 2))
  File "<ipython-input-39-b945b1917f3f>", line 15, in get_leaves
    for leaf in get_leaves(child_boards[i], depth - 1):
  File "<ipython-input-39-b945b1917f3f>", line 15, in get_leaves
    for leaf in get_leaves(child_boards[i], depth - 1):
  File "<ipython-input-39-b945b1917f3f>", line 15, in get_leaves
    for leaf in get_leaves(child_boards[i], depth - 1):
  [Previous line repeated 312 more times]
  File "<ipython-input-39-b945b1917f3f>", line 11, in get_leaves
    child_boards = [board.copy() for _ in legal_moves]
  File "<ipython-input-39-b945b1917f3f>", line 11, in <listcomp>
    child_boards = [board.copy() for _ in legal_moves]
  File "/Users/adam/anaconda/lib/python3.6/site-packages/

KeyboardInterrupt: 

In [ ]:
def get_leaf_nodes(self, node, leafs):
    if node is not None:
        if len(node.children) == 0:
            leafs.append(node)
        for n in node.children:
            self._collect_leaf_nodes(n, leafs)

In [56]:
16 * 1300/60/60

5.777777777777778

In [55]:
import time

gen = board_generator()
board = gen.__next__()

def get_leaves(board, depth, leaves):
    if depth == 0:
        leaves.append(board)
    else:
        legal_moves = list(board.legal_moves)
        child_boards = [board.copy() for _ in legal_moves]
        for i, move in enumerate(legal_moves):
            child_boards[i].push(legal_moves[i])
        for child_board in child_boards:
            get_leaves(child_board, depth-1, leaves)
    return leaves

t0 = time.time()
leaves = get_leaves(board, 3, [])
print(time.time() - t0)
print(len(leaves))

10.77158498764038
46398


depth 3, full tree no evaluation: 36782 nodes in 10 seconds, 15953 nodes in 16 seconds
depth 4, full tree no evaluation: 1309535 nodes in 511 seconds, negamax: 141934 nodes in 164 seconds

In [9]:
b = gen.__next__()
boards_and_moves = [(b.copy(), move) for move in b.legal_moves]
for board, move in boards_and_moves:
    board.push(move) 
boards = list(zip(*boards_and_moves))[0]
fvs = np.vstack([Chess.make_feature_vector(board) for board in boards])
values = sess.run(network.value, feed_dict={network.feature_vector_: fvs})
if b.turn:
    move_idx = np.argmax(values)
else:
    move_idx = np.argmin(values)

print(boards[move_idx], '\n\n', np.max(values))

env = Chess(b)
move, value = agent.get_move(sess, env, 1)
env.make_move(move)
print(env.board, '\n\n', value)
print(boards[move_idx] == env.board)

b . r r . b . .
p . . . . k . .
. . . . R . . p
. . p p . . B .
. . . . . . . .
. N P . . . . .
P . . . . P P P
. . . . R . K . 

 0.319721
b . r r . b . .
p . . . . k . .
. . . . R . . p
. . p p . . B .
. . . . . . . .
. N P . . . . .
P . . . . P P P
. . . . R . K . 

 0.319721
True


In [7]:
def make_feature_vector(board):
    piece_matrix = np.zeros((64, len(chess.PIECE_TYPES) + 1, len(chess.COLORS)))

    # piece positions
    for piece in chess.PIECE_TYPES:
        for color in chess.COLORS:
            piece_matrix[:, piece, int(color)] = ChessNeuralNetwork.pad_bitmask(board.pieces_mask(piece, color))

    # en passant target squares
    if board.ep_square:
        piece_matrix[board.ep_square, -1, int(board.turn)] = 1

    reshaped_piece_matrix = piece_matrix.reshape((64, (len(chess.PIECE_TYPES) + 1) * len(chess.COLORS)))
    feature_array = np.zeros((64, (len(chess.PIECE_TYPES) + 1) * len(chess.COLORS) + 2))
    feature_array[:, :-2] = reshaped_piece_matrix

    # empty squares
    empty_squares = (reshaped_piece_matrix.sum(axis=1) == 0)
    feature_array[empty_squares, :-2] = 1

    # castling rights
    feature_array[:, -1] = ChessNeuralNetwork.pad_bitmask(board.castling_rights)

    feature_vector = np.zeros((1, 1025))
    feature_vector[0, :-1] = np.reshape(feature_array, (1024,))
    feature_vector[0, -1] = board.turn

    return feature_vector

In [9]:
gen = board_generator()
board = gen.__next__()
feature_vector = ChessNeuralNetwork.make_feature_vector(board)
print(sess.run(network.value, feed_dict={network.feature_vector_: feature_vector})[0,0])
print(simple_value(board))

-0.197375
-0.197375320225


In [10]:
from random import choice, random

def generate_random_fen(density):
    pieces = ['p', 'n', 'b', 'r', 'q', 'P', 'N', 'B', 'R', 'Q']
    rank_fens = []
    for rank in range(8):
        rank_fen = ''
        empty_count = 0
        for file in range(8):
            if random() > density:
                empty_count += 1
            else:
                if empty_count > 0:
                    rank_fen += str(empty_count)
                    empty_count = 0
                piece = choice(pieces)
                rank_fen += piece
        if empty_count > 0:
            rank_fen += str(empty_count)
        rank_fens.append(rank_fen)
    fen = '/'.join(rank_fens)
    
    if random() < .5:
        fen += ' w '
    else:
        fen += ' b '
    casteling_rights = ''
    if random() < .5:
        casteling_rights += 'K'
    if random() < .5:
        casteling_rights += 'Q'
    if random() < .5:
        casteling_rights += 'k'
    if random() < .5:
        casteling_rights += 'q'
    if casteling_rights == '':
        casteling_rights = '--'
    fen += casteling_rights
        
    fen += ' - 0 1'
    return fen

In [11]:
def batch_generator(batch_size):
    while True:
        X = []
        y = []
        for i in range(batch_size):
            fen = generate_random_fen(random())
            board = chess.Board(fen)
            fv = ChessNeuralNetwork.make_feature_vector(board)
            X.append(fv)
            y.append(simple_value(board))
        yield np.array(X), np.array(y)

In [12]:
g = batch_generator(10)
X, y = g.__next__()
print(X)
print(y)

[[[ 0.  0.  0. ...,  0.  1.  1.]]

 [[ 0.  0.  0. ...,  0.  0.  1.]]

 [[ 0.  0.  0. ...,  0.  1.  0.]]

 ..., 
 [[ 0.  0.  0. ...,  0.  1.  0.]]

 [[ 0.  0.  1. ...,  0.  1.  0.]]

 [[ 0.  0.  0. ...,  0.  1.  1.]]]
[-0.99969858  0.9993293   0.99969858  0.37994896 -0.99986455  0.98367486
 -0.97574313 -0.53704957 -0.19737532  0.99263152]


In [22]:
fen = generate_random_fen(.1)
board = chess.Board(fen)
feature_vector = ChessNeuralNetwork.make_feature_vector(board)
print(sess.run(network.value, feed_dict={network.feature_vector_: feature_vector})[0, 0])
print(simple_value(board))

0.761594
0.761594155956


In [23]:
print(board)

. . p . . . . .
. . . . . . . .
. Q B . . . . .
. . . . . . . N
. . . . . . . .
. . . . . . . .
. . . . . q . .
. . . . . . . .


In [24]:
ChessNeuralNetwork.pad_bitmask(board.pieces_mask(2, 1)).reshape((8, 8))

array([[0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [1, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0]])

In [17]:
W_1 = np.zeros((1025, 100))
pieces = ['p', 'n', 'b', 'r', 'q', 'P', 'N', 'B', 'R', 'Q']
values = [-1, -3, -3, -5, -9, 1, 3, 3, 5, 9]
for piece, value in zip(pieces, values):
    fen = '/'.join([8*piece for _ in range(8)]) + ' b -- - 0 1'
    board = chess.Board(fen)
    W_1[ChessNeuralNetwork.make_feature_vector(board)[0] == 1, 0] = value

In [25]:
np.dot(feature_vector, W_1)

array([[ 5.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])

In [39]:
w = sess.run(network.trainable_variables)[0][:-1, :].reshape(64, -1, 100)\

In [46]:
w[:, 2, 0]

array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.], dtype=float32)

In [26]:
sess.run(network.value, feed_dict={network.feature_vector_: network.make_feature_vector(board)})

array([[ 0.76159418]], dtype=float32)

In [30]:
np.tanh(5/5)

0.76159415595576485

In [17]:
gen = board_generator()
board = gen.__next__()

In [18]:
%%timeit
foboard.legal_moves)

1000 loops, best of 3: 263 µs per loop


In [10]:
%%timeit
b = board.copy()
b.push(legal_moves[0])

10000 loops, best of 3: 106 µs per loop


In [ ]:
%%